coverage_point_calculator/src/speedtest.rs

```diff
@@ -0,0 +1,244 @@
+        Self(milliseconds)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Speedtests {
+    pub multiplier: Decimal,
+    pub speedtests: Vec<Speedtest>,
+}
+
+impl Speedtests {
+    pub fn new(speedtests: Vec<Speedtest>) -> Self {
+        // sort Newest to Oldest
+        let mut sorted_speedtests = speedtests;
+        sorted_speedtests.sort_by_key(|test| std::cmp::Reverse(test.timestamp));
+
+        let sorted_speedtests: Vec<_> = sorted_speedtests
+            .into_iter()
+            .take(MAX_ALLOWED_SPEEDTEST_SAMPLES)
+            .collect();
+
+        let multiplier = if sorted_speedtests.len() < MIN_REQUIRED_SPEEDTEST_SAMPLES {
+            SpeedtestTier::Fail.multiplier()
+        } else {
+            Speedtest::avg(&sorted_speedtests).multiplier()
+        };
+
+        Self {
+            multiplier,
+            speedtests: sorted_speedtests,
+        }
+    }
+
+    pub fn avg(&self) -> Speedtest {
+        Speedtest::avg(&self.speedtests)
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub struct Speedtest {
+    pub upload_speed: BytesPs,
+    pub download_speed: BytesPs,
+    pub latency: Millis,
+    pub timestamp: DateTime<Utc>,
+}
+
+impl Speedtest {
+    pub fn multiplier(&self) -> Decimal {
+        let upload = SpeedtestTier::from_upload(&self.upload_speed);
+        let download = SpeedtestTier::from_download(&self.download_speed);
+        let latency = SpeedtestTier::from_latency(&self.latency);
+
+        let tier = upload.min(download).min(latency);
+        tier.multiplier()
+    }
+
+    pub fn avg(speedtests: &[Self]) -> Self {
+        let mut download = 0;
+        let mut upload = 0;
+        let mut latency = 0;
+
+        for test in speedtests {
+            upload += test.upload_speed.0;
+            download += test.download_speed.0;
+            latency += test.latency.0;
+        }
+
+        let count = speedtests.len();
+        Self {
+            upload_speed: BytesPs::new(upload / count as u64),
+            download_speed: BytesPs::new(download / count as u64),
+            latency: Millis::new(latency / count as u32),
+            timestamp: Utc::now(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SpeedtestTier {
+    Good = 4,
+    Acceptable = 3,
+    Degraded = 2,
+    Poor = 1,
+    Fail = 0,
+}
+
+impl SpeedtestTier {
+    pub fn multiplier(&self) -> Decimal {
+        match self {
+            SpeedtestTier::Good => dec!(1.00),
+            SpeedtestTier::Acceptable => dec!(0.75),
+            SpeedtestTier::Degraded => dec!(0.50),
+            SpeedtestTier::Poor => dec!(0.25),
+            SpeedtestTier::Fail => dec!(0),
+        }
+    }
+
+    fn from_download(bytes: &BytesPs) -> Self {
+        match bytes.as_mbps() {
+            100.. => Self::Good,
+            75.. => Self::Acceptable,
+            50.. => Self::Degraded,
+            30.. => Self::Poor,
+            _ => Self::Fail,
+        }
```

Comment:
personally, I think this is a little easier to read as just if/else, but I don't really have a horse in the race. Just consider it:
```suggestion
        let mbps = bytes.as_mbps();
        if mbps >= 100 {
            Self::Good
        } else if mbps >= 75 {
            Self::Acceptable
        } else if mbps >= 50 {
            Self::Degraded
        } else if mbps >= 30 {
            Self::Poor
        } else {
            Self::Fail
        }
```

Reply:
Considered. Maybe one day I'll come around, but tis not this day.